mlflow/gateway/providers/mosaicml.py

```diff
@@ -27,6 +31,37 @@ async def _request(self, model: str, payload: Dict[str, Any]) -> Dict[str, Any]:
             payload=payload,
         )
 
+    @staticmethod
+    def _get_used_tokens_estimate(payload):
+        """
+        Computes the token count estimate for a given payload.
+        """
+        if isinstance(payload, chat.RequestPayload):
+            return sum(len(msg.content.split()) for msg in payload.messages)
+        elif isinstance(payload, completions.RequestPayload):
+            return len(payload.prompt.split())
+        return 0
+
+    # NB: This validation logic can be removed iff the token allocation validation error moves from
+    # a 5xx error to a 4xx error.
+    def _is_token_count_over_threshold(
+        self, payload: Union[chat.RequestPayload, completions.RequestPayload]
+    ):
+        """
+        Performs a conservative estimate of submitted token counts and a requested max_tokens value
+        to preempt a server-side retry based on a 5xx return exception from MosaicML for token
+        count violating the maximum threshold.
+        Since we don't have access to the tokenizer that is used for a given model, the
+        conservative estimation of user-provided tokens errs on the side of caution by counting
+        each word supplied as consuming 2 tokens.
+        """
+        used_tokens = self._get_used_tokens_estimate(payload)
+        max_tokens = payload.max_tokens or 0
+        return (
```

Comment:
Hmm I'm not sure if this is the best temporary workaround since we may still run into the same error when the estimate isn't conservative enough, but in most cases, when it's too conservative, we throw prematurely. It might be better actually to just parse the error when sending the request to MosaicML and check to see if the text contains `prompt token count + max output tokens cannot exceed 4096`

Reply:
Yeah, definitely going to just dig in and convert the response error directly from the aiohttp request. 
Should have a rework of this solution soon :)